management/server/peer.go

```diff
@@ -574,19 +583,51 @@ func (am *DefaultAccountManager) LoginPeer(login PeerLogin) (*nbpeer.Peer, *Netw
 		return nil, nil, status.Errorf(status.Internal, "failed while logging in peer")
 	}
 
-	// we found the peer, and we follow a normal login flow
-	unlock := am.Store.AcquireAccountWriteLock(account.Id)
-	defer unlock()
+	peer, err := am.Store.GetPeerByPeerPubKey(login.WireGuardPubKey)
+	if err != nil {
+		return nil, nil, status.NewPeerNotRegisteredError()
+	}
+
+	accSettings, err := am.Store.GetAccountSettings(accountID)
+	if err != nil {
+		return nil, nil, status.Errorf(status.Internal, "failed to get account settings: %s", err)
+	}
+
+	// duplicated logic from after the lock to have an early exit
+	expired := peerLoginExpired(peer, accSettings)
+	if expired {
+		err = checkAuth(login.UserID, peer)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	var isWriteLock bool
+	var unlock func()
+	// early detection of what type of lock we need
+	if peer.UpdateMetaIfNew(login.Meta) || expired {
+		unlock = am.Store.AcquireAccountWriteLock(accountID)
+		isWriteLock = true
+	} else {
+		unlock = am.Store.AcquireAccountReadLock(accountID)
+		isWriteLock = false
+	}
+	log.Debugf("peer changed meta or login expired, acquiring write lock: %v", isWriteLock)
```

Comment:
this log should be in line 609

Reply:
No, the log is true/false so it will be logged either way. I could do 2 logs then it is mor explicit